### PR TITLE
CMakeLists.txt: Check the Windows SDK version

### PR DESCRIPTION
**Pull request by StephanTLavavej**
Toolset update #2714 began requiring the Windows 11 SDK 22000 to build and test the STL. Although it's mentioned in the README, if a contributor misses this dependency, the resulting error message is incomprehensible (as @miscco discovered). @strega-nil-ms suggested adding a build check.

Fortunately, the Windows SDK provides a macro to identify itself, `WDK_NTDDI_VERSION`. (Unlike `NTDDI_VERSION`, this is not subject to user overrides.) It's present in the default WinSDK, where it was set to "Vibranium", whereas in the latest WinSDK, it's set to "Cobalt":

* `C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\shared\sdkddkver.h`
  + `#define WDK_NTDDI_VERSION                   NTDDI_WIN10_VB /* ABRACADABRA_WIN10_VB */`
* `C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\shared\sdkddkver.h`
  + `#define WDK_NTDDI_VERSION                   NTDDI_WIN10_CO`

I've tested this by changing my installed WinSDK. Here's the user experience:

### :x: Windows 10 SDK 19041, failure:

```
C:\GitHub\STL>cmake -G Ninja -S . -B out\build\x64
-- The CXX compiler identification is MSVC 19.33.31424.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Preview/VC/Tools/MSVC/14.33.31424/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test WINDOWS_SDK_VERSION_CHECK
-- Performing Test WINDOWS_SDK_VERSION_CHECK - Failed
CMake Error at CMakeLists.txt:15 (message):
  The STL must be built with the Windows 11 SDK (10.0.22000.0) or later.
  Make sure it's available by selecting it in the Individual Components tab
  of the VS Installer.


-- Configuring incomplete, errors occurred!
See also "C:/GitHub/STL/out/build/x64/CMakeFiles/CMakeOutput.log".
See also "C:/GitHub/STL/out/build/x64/CMakeFiles/CMakeError.log".
```

### :white_check_mark: Windows 11 SDK 22000, success:

```
C:\GitHub\STL>cmake -G Ninja -S . -B out\build\x64
-- The CXX compiler identification is MSVC 19.33.31424.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Preview/VC/Tools/MSVC/14.33.31424/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test WINDOWS_SDK_VERSION_CHECK
-- Performing Test WINDOWS_SDK_VERSION_CHECK - Success
-- Searching for VS clang-format
-- Searching for VS clang-format - found
-- Boost.Math: standalone mode ON
-- Found Python: C:/Users/stl/AppData/Local/Programs/Python/Python310/python.exe (found suitable version "3.10.4", minimum required is "3.9") found components: Interpreter
-- Configuring done
-- Generating done
-- Build files have been written to: C:/GitHub/STL/out/build/x64
```

### CODE REVIEW

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,16 @@ cmake_minimum_required(VERSION 3.22)
 set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
 project(msvc_standard_libraries LANGUAGES CXX)
 
+include(CheckCXXSourceCompiles)
+check_cxx_source_compiles("
+#include <sdkddkver.h>
+static_assert(WDK_NTDDI_VERSION >= NTDDI_WIN10_CO, \"Inspecting WDK_NTDDI_VERSION, the Windows SDK version.\");
+int main() {}
+" WINDOWS_SDK_VERSION_CHECK)
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply CMakeLists.txt]: Awesome, I didn't know about this!

File: CMakeLists.txt [hunk 2]
```diff
@@ -5,6 +5,16 @@ cmake_minimum_required(VERSION 3.22)
 set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
 project(msvc_standard_libraries LANGUAGES CXX)
 
+include(CheckCXXSourceCompiles)
```
Comment (bhardwajs) [inline CMakeLists.txt:8]: Curious why this approach? Wouldn't `find_package(WindowsSDK)` work? And, it also can tell which directory was found etc.?

https://github.com/rpavlik/cmake-modules/blob/main/FindWindowsSDK.cmake

Comment (strega-nil) [reply CMakeLists.txt:8]: This would be depending on a full find module for, in my opinion, no real reason. This works and gives us the result we want.

File: CMakeLists.txt [hunk 3]
```diff
@@ -5,6 +5,16 @@ cmake_minimum_required(VERSION 3.22)
...
 project(msvc_standard_libraries LANGUAGES CXX)
 
+include(CheckCXXSourceCompiles)
+check_cxx_source_compiles([=[
+#include <sdkddkver.h>
+static_assert(WDK_NTDDI_VERSION >= NTDDI_WIN10_CO, "Inspecting WDK_NTDDI_VERSION, the Windows SDK version.");
+int main() {}
+]=] WINDOWS_SDK_VERSION_CHECK)
+if(NOT WINDOWS_SDK_VERSION_CHECK)
+    message(FATAL_ERROR "To build the STL, select the Windows 11 SDK (10.0.22000.0) in the VS Installer.")
```
Comment (strega-nil) [inline CMakeLists.txt:15]: nit:

```suggestion
    message(FATAL_ERROR "The STL must be built with the Windows 11 SDK or later. Make sure it's available by installing the Windows 11 SDK (at least version 10.0.22000.0) in the VS Installer.")
```

I think this error message is more obvious.

**Review by strega-nil — APPROVED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (barcharcraz) [reply CMakeLists.txt:8]: Yeah, it would be one thing if it was a builtin module, but it looks like it is not

File: CMakeLists.txt (hunk 3 above)
Comment (StephanTLavavej) [reply CMakeLists.txt:15]: Hmm, I agree that the additional explanation is an improvement. However, I'd like to rephrase it to be future-proof - currently we require the first version of the Windows 11 SDK, but there will be more, and the message should remain accurate if we just increase the number. Will push a change...

**Review by barcharcraz — APPROVED** [review]
OK I can't think of a better way to do this, I don't like the compiler checks though, as they are slow.

**Comment by StephanTLavavej** [comment]
Regarding slowness, I measured the cost with:
```bat
"C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Auxiliary\Build\vcvars64.bat"
pushd C:\GitHub\STL
git clean -x -d -f
pwsh
```
```ps1
Measure-Command { cmake -G Ninja -S . -B out\build\x64 | Out-Default }
```
I compared `main` and this branch, and averaged 3 runs. (Note that this excludes the time to clean etc.)

* `main` took 3779 ms on average (raw times 3796.2115, 3731.3771, 3809.0518).
* `winsdk-check` took 3924 ms on average (raw times 3930.953, 3918.3531, 3921.5556).

Thus this costs **~145 ms** on my i7-10700.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.
